#108240 Rename 'src/bootstrap/native.rs' to llvm.rs

rohaquinlop opened the pull request:
Fixed #108240

Renamed 'native.rs' to 'llvm.rs', also moved `TestHelpers` to `test.rs`. Replaced all the `native.rs` occurrences at `src/bootstrap` files to `llvm.rs`

Jules-Bertholet committed:
Add `round_ties_even` to `f32` and `f64`

cjgillot committed:
Introduce a no-op PlaceMention statement for `let _ =`.

davidtwco committed:
various: translation resources from cg backend

Extend `CodegenBackend` trait with a function returning the translation
resources from the codegen backend, which can be added to the complete
list of resources provided to the emitter.

Signed-off-by: David Wood <david.wood@huawei.com>

Jules-Bertholet committed:
Use rint instead of roundeven

Use rint intrinsic instead of roundeven to impement `round_ties_even`. They do the same thing when rounding mode is default, which Rust assumes.
And `rint` has better platform support.

Keeps `roundeven` around in `core::intrinsics`, it's doing no harm there.

emilyalbini committed:
replace legacy copyright annotations in submodules

aliemjay committed:
remove obsolete `givens` from regionck

aliemjay committed:
resolve regions before implied bounds

aliemjay committed:
add test

bjorn3 committed:
Update Cranelift to 0.93.0

Closes #1352

bjorn3 committed:
Use new abi for 128bit int to float cast intrinsics on Windows

bjorn3 committed:
Avoid spurious visible_parent_map query invocation

bjorn3 committed:
Remove a couple of duplicate layout_of and monomorphize calls

bjorn3 committed:
Merge commit '7d53619064ab7045c383644cb445052d2a3d46db' into sync_cg_clif-2023-02-09

bjorn3 committed:
Merge branch 'sync_from_rust'

eggyal committed:
Alias folding/visiting traits instead of re-export

matthiaskrgr committed:
clippy::perf fixes

matthiaskrgr committed:
clippy::complexity fixes

bjorn3 committed:
Fix signaling available_token_condvar when a new token is received

This slightly improves performance on systems with many cores and
barely affects systems with few cores.

My laptop (2 core + HT):
Before:
Benchmark 2: RUSTC=rustc /home/bjorn/Projects/cg_clif2/./dist/cargo-clif build --manifest-path /home/bjorn/Projects/cg_clif2/./download/simple-raytracer/Cargo.toml --target-dir /home/bjorn/Projects/cg_clif2/./build/simple_raytracer
  Time (mean ± σ):     12.042 s ±  0.313 s    [User: 29.434 s, System: 4.720 s]
  Range (min … max):   11.670 s … 12.795 s    10 runs

After:
Benchmark 2: RUSTC=rustc /home/bjorn/Projects/cg_clif2/./dist/cargo-clif build --manifest-path /home/bjorn/Projects/cg_clif2/./download/simple-raytracer/Cargo.toml --target-dir /home/bjorn/Projects/cg_clif2/./build/simple_raytracer
  Time (mean ± σ):     12.037 s ±  0.384 s    [User: 29.960 s, System: 4.722 s]
  Range (min … max):   11.673 s … 12.769 s    10 runs

Dev desktop (32 cores)
Before:
Benchmark 2: RUSTC=rustc /home/gh-bjorn3/cg_clif/./dist/cargo-clif build --manifest-path /home/gh-bjorn3/cg_clif/./download/simple-raytracer/Cargo.toml --target-dir /home/gh-bjorn3/cg_clif/./build/simple_raytracer
  Time (mean ± σ):     10.425 s ±  0.104 s    [User: 25.877 s, System: 5.513 s]
  Range (min … max):   10.267 s … 10.640 s    10 runs

After:
Benchmark 2: RUSTC=rustc /home/gh-bjorn3/cg_clif/./dist/cargo-clif build --manifest-path /home/gh-bjorn3/cg_clif/./download/simple-raytracer/Cargo.toml --target-dir /home/gh-bjorn3/cg_clif/./build/simple_raytracer
  Time (mean ± σ):     10.212 s ±  0.100 s    [User: 25.918 s, System: 5.555 s]
  Range (min … max):   10.079 s … 10.362 s    10 runs

bjorn3 committed:
Merge pull request #1355 from matthiaskrgr/clippy_perf

clippy::perf fixes

bjorn3 committed:
Merge pull request #1356 from matthiaskrgr/clippy_compl

clippy::complexity fixes

cjgillot committed:
Implement checked Shl/Shr at MIR building.

chbaker0 committed:
Don't export of __rust_* alloc symbols if not codegened

compiler-errors committed:
Make permit_uninit/zero_init fallible

oli-obk committed:
s/eval_usize/eval_target_usize/ for clarity

bjorn3 committed:
Sync from rust e9ab7872fd77861e3d182ea85a82c4649c5bb3f8

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (065852def 2023-02-13)

bjorn3 committed:
Fix non-dynamic indexing into vector types

WaffleLapkin committed:
`if $c:expr { Some($r:expr) } else { None }` =>> `$c.then(|| $r)`

cjgillot committed:
Adapt cg_clif.

bors committed:
Auto merge of #108012 - compiler-errors:issue-107999, r=oli-obk

Don't ICE in `might_permit_raw_init` if reference is polymorphic

Emitting optimized MIR for a polymorphic function may require computing layout of a type that isn't (yet) known. This happens in the instcombine pass, for example. Let's fail gracefully in that condition.

cc `@saethlin`
fixes #107999

nnethercote committed:
Replace `mk_foo` calls with `infer_foo` where possible.

There are several `mk_foo`/`intern_foo` pairs, where the former takes an
iterator and the latter takes a slice. (This naming convention is bad,
but that's a fix for another PR.)

This commit changes several `mk_foo` occurrences into `intern_foo`,
avoiding the need for some `.iter()`/`.into_iter()` calls. Affected
cases:
- mk_type_list
- mk_tup
- mk_substs
- mk_const_list

bjorn3 committed:
Remove easy_call

bjorn3 committed:
Add fixme

nnethercote committed:
Rename many interner functions.

(This is a large commit. The changes to
`compiler/rustc_middle/src/ty/context.rs` are the most important ones.)

The current naming scheme is a mess, with a mix of `_intern_`, `intern_`
and `mk_` prefixes, with little consistency. In particular, in many
cases it's easy to use an iterator interner when a (preferable) slice
interner is available.

The guiding principles of the new naming system:
- No `_intern_` prefixes.
- The `intern_` prefix is for internal operations.
- The `mk_` prefix is for external operations.
- For cases where there is a slice interner and an iterator interner,
  the former is `mk_foo` and the latter is `mk_foo_from_iter`.

Also, `slice_interners!` and `direct_interners!` can now be `pub` or
non-`pub`, which helps enforce the internal/external operations
division.

It's not perfect, but I think it's a clear improvement.

The following lists show everything that was renamed.

slice_interners
- const_list
  - mk_const_list -> mk_const_list_from_iter
  - intern_const_list -> mk_const_list
- substs
  - mk_substs -> mk_substs_from_iter
  - intern_substs -> mk_substs
  - check_substs -> check_and_mk_substs (this is a weird one)
- canonical_var_infos
  - intern_canonical_var_infos -> mk_canonical_var_infos
- poly_existential_predicates
  - mk_poly_existential_predicates -> mk_poly_existential_predicates_from_iter
  - intern_poly_existential_predicates -> mk_poly_existential_predicates
  - _intern_poly_existential_predicates -> intern_poly_existential_predicates
- predicates
  - mk_predicates -> mk_predicates_from_iter
  - intern_predicates -> mk_predicates
  - _intern_predicates -> intern_predicates
- projs
  - intern_projs -> mk_projs
- place_elems
  - mk_place_elems -> mk_place_elems_from_iter
  - intern_place_elems -> mk_place_elems
- bound_variable_kinds
  - mk_bound_variable_kinds -> mk_bound_variable_kinds_from_iter
  - intern_bound_variable_kinds -> mk_bound_variable_kinds

direct_interners
- region
  - intern_region (unchanged)
- const
  - mk_const_internal -> intern_const
- const_allocation
  - intern_const_alloc -> mk_const_alloc
- layout
  - intern_layout -> mk_layout
- adt_def
  - intern_adt_def -> mk_adt_def_from_data (unusual case, hard to avoid)
  - alloc_adt_def(!) -> mk_adt_def
- external_constraints
  - intern_external_constraints -> mk_external_constraints

Other
- type_list
  - mk_type_list -> mk_type_list_from_iter
  - intern_type_list -> mk_type_list
- tup
  - mk_tup -> mk_tup_from_iter
  - intern_tup -> mk_tup

bjorn3 committed:
Move windows i128 argument by-ref handling to lib_call

bjorn3 committed:
Move windows i128 indirect return handling to lib_call

bjorn3 committed:
Sync from rust 0416b1a6f

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (2d14db321 2023-02-15)

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (0416b1a6f 2023-02-14)

bjorn3 committed:
Sync from rust 9a7cc6c32f1a690f86827e4724bcda85e506ef35

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (9a7cc6c32 2023-02-16)

bjorn3 committed:
Support updating to specific dates in rustup.sh

Fixes #1294

bjorn3 committed:
Fix float to int compiler builtin call abi

bjorn3 committed:
Update to compiler-builtins 0.1.87

bjorn3 committed:
Split out lib_call_unadjusted function

bjorn3 committed:
Fix return address type

bjorn3 committed:
Check output of checked_div in std_example

bjorn3 committed:
Remove incorrect check in lib_call

bjorn3 committed:
Fix abi for checked multiplication

bjorn3 committed:
Merge pull request #1359 from bjorn3/fix_llvm14-builtins-abi

Fix compiler-builtins 0.1.87 on Windows

cjgillot committed:
Stop implementing _with_overflow intrinsics in codegen backends.

bjorn3 committed:
Sync from rust 4507fdaaa27ea2fb59a41df2ce7d1f290da53dae

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (4507fdaaa 2023-02-18)

bjorn3 committed:
Add wrappers to dist/bin/ too in addition to dist/

Rust's build system since recently expects rustc to be in a bin/
directory if it is specified using the rustc option in config.toml.

bjorn3 committed:
Enable inline stack probes on riscv64

Fixes #1358

bjorn3 committed:
Fix warnings

bjorn3 committed:
Allow codegen to unsize dyn* to dyn

cg_clif counterpart to rust-lang/rust#106532

bjorn3 committed:
Handle wrappers around dyn* in receiver types

Fixes #1324

bjorn3 committed:
Update portable-simd

bjorn3 committed:
Add test for #1318

bjorn3 committed:
Fix typo

bjorn3 committed:
Move abi-cafe runs to a separate workflow

bjorn3 committed:
Only allow a single abi-cafe run at a time per branch

bjorn3 committed:
Fix warning

bjorn3 committed:
Rename build job to test

bjorn3 committed:
Sync from rust 7aa413d59206fd511137728df3d9e0fd377429bd

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (7aa413d59 2023-02-19)

cjgillot committed:
Update MIR doc.

eggyal committed:
Remove type-traversal trait aliases

bors committed:
Auto merge of #108340 - eggyal:remove_traversal_trait_aliases, r=oli-obk

Remove type-traversal trait aliases

#107924 moved the type traversal (folding and visiting) traits into the type library, but created trait aliases in `rustc_middle` to minimise both the API churn for trait consumers and the arising boilerplate.  As mentioned in that PR, an alternative approach of defining subtraits with blanket implementations of the respective supertraits was also considered at that time but was ruled out as not adding much value.

Unfortunately, it has since emerged that rust-analyzer has difficulty with these trait aliases at present, resulting in a degraded contributor experience (see the recent [r-a has become useless](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/r-a.20has.20become.20useless) topic on the #t-compiler/help Zulip stream).

This PR removes the trait aliases, and accordingly the underlying type library traits are now used directly; they are parameterised by `TyCtxt<'tcx>` rather than just the `'tcx` lifetime, and imports have been updated to reflect the fact that the trait aliases' explicitly named traits are no longer automatically brought into scope.  These changes also roll-back the (no-longer required) workarounds to #107747 that were made in b409329c624b9e3bbd7d8e07697e2e9f861a45b6.

Since this PR is just a find+replace together with the changes necessary for compilation & tidy to pass, it's currently just one mega-commit.  Let me know if you'd like it broken up.

r? `@oli-obk`

Noratrieb committed:
Unify validity checks into a single query

Previously, there were two queries to check whether a type allows the
0x01 or zeroed bitpattern.

I am planning on adding a further initness to check, truly uninit for
MaybeUninit, which would make this three queries. This seems overkill
for such a small feature, so this PR unifies them into one.

bjorn3 committed:
Sync from rust c5c7d2b37780dac1092e75f12ab97dd56c30861d

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (c5c7d2b37 2023-02-24)

bjorn3 committed:
Reduce verbosity of libcore testing

bjorn3 committed:
Silence almost all warnings when compiling coretests

bjorn3 committed:
Reduce verbosity of rand testing

bjorn3 committed:
Patch coretests separately from the standard library source

bjorn3 committed:
Sync from rust 34e6673a0473e90ef01a18eb575392c9e3859747

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (34e6673a0 2023-02-25)

bjorn3 committed:
Re-enable DataflowConstProp for standard library builds

bjorn3 committed:
Update test repo versions

bjorn3 committed:
Remove --cap-lint warn from regex test

bjorn3 committed:
Remove no longer necessary code removing "[codegen mono items]" from test output

bjorn3 committed:
Allow multiple threads and panicking tests when testing regex

I guess this was a leftover from very early in the development of
cg_clif. Allowing multiple threads significantly improves performance,
while panicking tests can run now thanks to -Zpanic-abort-tests.

bjorn3 committed:
Run tests for all crates in the regex workspace

bjorn3 committed:
Avoid a duplicate "[TEST] rust-random/rand"

bjorn3 committed:
Merge branch 'staging'

Noratrieb committed:
Unify all validity check intrinsics

Also merges the inhabitedness check into the query to further unify the
code paths.

matthiaskrgr committed:
Rollup merge of #108364 - Nilstrieb:validity-checks-refactor, r=compiler-errors

Unify validity checks into a single query

Previously, there were two queries to check whether a type allows the 0x01 or zeroed bitpattern.

I am planning on adding a further initness to check in #100423, truly uninit for MaybeUninit, which would make this three queries. This seems overkill for such a small feature, so this PR unifies them into one.

I am not entirely happy with the naming and key type and open for improvements.

r? oli-obk

clubby789 committed:
Remove uses of `box_syntax` in rustc and tools

bjorn3 committed:
Fix loading of dylibs not in the search path in jit mode

Reported in #1249

petrhosek committed:
Support for Fuchsia RISC-V target

Fuchsia is in the process of implementing the RISC-V support. This
change implements the minimal Rust compiler support. The support for
building runtime libraries will be implemented in follow up changes
once Fuchsia SDK has the RISC-V support.

bjorn3 committed:
Sync from rust 44cfafe2fafe816395d3acc434663a45d5178c41

bjorn3 committed:
Rustup to rustc 1.69.0-nightly (44cfafe2f 2023-03-03)

bjorn3 committed:
Rustfmt

bjorn3 committed:
Fix rustc test suite

bjorn3 committed:
Remove copy_nonoverlapping intrinsic definition

It has been getting lowered to a dedicated MIR statement for a while

zeegomo committed:
Remove DropAndReplace terminator

PR 107844 made DropAndReplace unused, let's remove it completely
from the codebase.

lqd committed:
fix ignore header in MSVC test

bors committed:
Auto merge of #95317 - Jules-Bertholet:round_ties_to_even, r=pnkfelix,m-ou-se,scottmcm

Add `round_ties_even` to `f32` and `f64`

Tracking issue: #96710

Redux of #82273. See also #55107

Adds a new method, `round_ties_even`, to `f32` and `f64`, that rounds the float to the nearest integer , rounding halfway cases to the number with an even least significant bit. Uses the `roundeven` LLVM intrinsic to do this.

Of the five IEEE 754 rounding modes, this is the only one that doesn't already have a round-to-integer function exposed by Rust (others are `round`, `floor`, `ceil`, and `trunc`).  Ties-to-even is also the rounding mode used for int-to-float and float-to-float `as` casts, as well as float arithmentic operations. So not having an explicit rounding method for it seems like an oversight.

Bikeshed: this PR currently uses `round_ties_even` for the name of the method. But maybe `round_ties_to_even` is better, or `round_even`, or `round_to_even`?

spastorino committed:
Filter out RPITITs in astconv when checking for missing associated types

spastorino committed:
Filter out RPITITs in object_safety_violations_for_trait

spastorino committed:
Run existing async in traits tests using -Zlower-impl-trait-in-trait-to-assoc-ty

spastorino committed:
Remove tests/ui/impl-trait/in-trait/new-lowering-strategy in favor of using revisions on existing tests

spastorino committed:
Run existing impl trait in traits tests using -Zlower-impl-trait-in-trait-to-assoc-ty

spastorino committed:
Fix object safety checks for new RPITITs

bjorn3 committed:
Update Cranelift to 0.93.1

This fixes a potential miscompilation

matthiaskrgr committed:
Rollup merge of #108856 - Zeegomo:remove-drop-and-rep, r=tmiasko

Remove DropAndReplace terminator

#107844 made DropAndReplace unused, let's remove it completely from the codebase.

spastorino committed:
Get impl defaultness using query

spastorino committed:
Make fns from other crates with RPITIT work

spastorino committed:
Remove some direct calls to local_def_id_to_hir_id on diagnostics

aliemjay committed:
exhaustive match on implied bounds regions

bjorn3 committed:
Sync from rust 39f2657d1101b50f9b71ae460b762d330cc8426b

bjorn3 committed:
Rustup to rustc 1.70.0-nightly (39f2657d1 2023-03-09)

bjorn3 committed:
Fix rustc test suite

onur-ozkan committed:
add `enable-warnings` flag for llvm

Signed-off-by: ozkanonur <work@onurozkan.dev>

matthiaskrgr committed:
Rollup merge of #108017 - chbaker0:fix-105967, r=chbaker0

Add `--no-undefined-version` link flag and fix associated breakage

LLVM upstream sets `--no-undefined-version` by default in lld: https://reviews.llvm.org/D135402.

Due to a bug in how version scripts are generated, this breaks the `dylib` output type for most crates. See https://github.com/rust-lang/rust/issues/105967#issuecomment-1428671533 for details.

This PR adds the flag to gcc flavor linkers in anticipation of this LLVM change rolling in, and patches `rustc` to not attempt to export `__rust_*` allocator symbols when they weren't generated.

Fixes #105967

bors committed:
Auto merge of #104527 - ferrocene:pa-more-licenses, r=pnkfelix

Add more license annotations

This PR updates the `.reuse/dep5` file to include more accurate licensing data for everything in the repository (*excluding* submodules and dependencies). Some decisions were made in this PR:

* The standard copyright attribution for files maintained by us is "The Rust Project Developers (see https://thanks.rust-lang.org)", to avoid having to maintain an in-tree `AUTHORS` file.
* For files that have specific licensing terms, we added the terms to the `.reuse/dep5` rather than adding SPDX comments in the files themselves.
* REUSE picks up any comment/text line with `Copyright` on it, so I had to sprinkle around `REUSE-IgnoreStart` and `REUSE-IgnoreEnd` comments.

The rendered `COPYRIGHT` file is available at https://gist.github.com/pietroalbini/efb81103f69596d39758114f3f6a8688.

r? `@pnkfelix`

bors committed:
Auto merge of #109001 - matthiaskrgr:rollup-a3agnwp, r=matthiaskrgr

Rollup of 8 pull requests

Successful merges:

 - #105798 (Relax ordering rules for `asm!` operands)
 - #105962 (Stabilize path_as_mut_os_str)
 - #106085 (use problem matchers for tidy CI)
 - #107711 (Stabilize movbe target feature)
 - #108017 (Add `--no-undefined-version` link flag and fix associated breakage)
 - #108891 (Remove an extraneous include)
 - #108902 (no more do while :<)
 - #108912 (Document tool lints)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

scottmcm committed:
`MaybeUninit::assume_init_read` should have `noundef` load metadata

I was looking into `array::IntoIter` optimization, and noticed that it wasn't annotating the loads with `noundef` for simple things like `array::IntoIter<i32, N>`.

Turned out to be a more general problem as `MaybeUninit::assume_init_read` isn't marking the load as initialized (<https://rust.godbolt.org/z/Mxd8TPTnv>), which is unfortunate since that's basically its reason to exist.

This PR lowers `ptr::read(p)` to `copy *p` in MIR, which fortuitiously also improves the IR we give to LLVM for things like `mem::replace`.

scottmcm committed:
Add a codegen test to confirm this fixes 106369

scottmcm committed:
Add a codegen test to confirm this fixes 73258

scottmcm committed:
Add a better approach comment in `ptr::read` to justify the intrinsic

bjorn3 committed:
Sync from rust 7b4f48927dce585f747a58083b45ab62b9d73a53

bjorn3 committed:
Rustup to rustc 1.70.0-nightly (7b4f48927 2023-03-12)

lqd committed:
fix error pattern

lqd committed:
remove need for test expectations

compiler-errors committed:
Encode opt_rpitit_info for associated types

compiler-errors committed:
Layout of `&dyn Trait<[type error]>` is still wide

compiler-errors committed:
Don't ICE for late-bound consts across AnonConstBoundary

compiler-errors committed:
Use `unused_generic_params` from crate metadata

compiler-errors committed:
Don't codegen impossible to satisfy impls

mu001999 committed:
Create dir for build_triple

MaciejWas committed:
Emit "modifies receiver" diagnostic when no method is found

If no method is found when checking method call, we check  if we called a method with signature (&mut T, ...) -> (). If this is the case then we emit a diagnostic message

matthiaskrgr committed:
Rollup merge of #108880 - spastorino:new-rpitit-6, r=compiler-errors

Remove tests/ui/impl-trait/in-trait/new-lowering-strategy in favor of using revisions on existing tests

r? `@compiler-errors`

This one again sits on top of existing approved PRs and it still needs to add revisions to tests in `tests/ui/impl-trait/in-trait` as it only does so for async in traits.

matthiaskrgr committed:
Rollup merge of #108722 - petrhosek:fuchsia-riscv, r=petrochenkov

Support for Fuchsia RISC-V target

Fuchsia is in the process of implementing the RISC-V support. This change implements the minimal Rust compiler support. The support for building runtime libraries will be implemented in follow up changes once Fuchsia SDK has the RISC-V support.

matthiaskrgr committed:
Rollup merge of #108909 - spastorino:new-rpitit-7, r=compiler-errors

Fix object safety checks for new RPITITs

This one goes on top of #108869

r? `@compiler-errors`

matthiaskrgr committed:
Rollup merge of #108915 - spastorino:new-rpitit-8, r=compiler-errors

Remove some direct calls to local_def_id_to_hir_id on diagnostics

Was playing with `tests/ui/impl-trait/in-trait/default-body-with-rpit.rs` and was able to remove some ICEs. Still getting ...

```
error[E0277]: `impl Future<Output = Foo::{opaque#0}>` is not a future
  --> tests/ui/impl-trait/in-trait/default-body-with-rpit.rs:10:28
   |
10 |     async fn baz(&self) -> impl Debug {
   |                            ^^^^^^^^^^ `impl Future<Output = Foo::{opaque#0}>` is not a future
   |
   = help: the trait `Future` is not implemented for `impl Future<Output = Foo::{opaque#0}>`
   = note: impl Future<Output = Foo::{opaque#0}> must be a future or must implement `IntoFuture` to be awaited
note: required by a bound in `Foo::{opaque#1}`
  --> tests/ui/impl-trait/in-trait/default-body-with-rpit.rs:10:28
   |
10 |     async fn baz(&self) -> impl Debug {
   |                            ^^^^^^^^^^ required by this bound in `Foo::{opaque#1}`

error[E0277]: the size for values of type `impl Future<Output = Foo::{opaque#0}>` cannot be known at compilation time
  --> tests/ui/impl-trait/in-trait/default-body-with-rpit.rs:10:28
   |
10 |     async fn baz(&self) -> impl Debug {
   |                            ^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `impl Future<Output = Foo::{opaque#0}>`
note: required by a bound in `Foo::{opaque#1}`
  --> tests/ui/impl-trait/in-trait/default-body-with-rpit.rs:10:28
   |
10 |     async fn baz(&self) -> impl Debug {
   |                            ^^^^^^^^^^ required by this bound in `Foo::{opaque#1}`

error: internal compiler error: compiler/rustc_hir_typeck/src/closure.rs:724:18: async fn generator return type not an inference variable: Foo::{opaque#1}<'_>
  --> tests/ui/impl-trait/in-trait/default-body-with-rpit.rs:10:39
   |
10 |       async fn baz(&self) -> impl Debug {
   |  _______________________________________^
11 | |         ""
12 | |     }
   | |_____^
```

But I guess this is a little bit of progress anyway.

This one goes on top of #108700 and #108945
r? `@compiler-errors`

matthiaskrgr committed:
Rollup merge of #109101 - compiler-errors:layout-err, r=michaelwoerister

Fall back to old metadata computation when type references errors

Projection is a bit too aggressive normalizing `<dyn Trait<[type error]> as Pointee>::Metadata` to `[type error]`, rather than to `DynMetadata<..>`. Side-step that by just falling back to the old structural metadata computation.

Fixes #109078

matthiaskrgr committed:
Rollup merge of #108923 - spastorino:new-rpitit-9, r=compiler-errors

Make fns from other crates with RPITIT work for -Zlower-impl-trait-in-trait-to-assoc-ty

Only the last two commits are meaningful.

r? `@compiler-errors`

matthiaskrgr committed:
Rollup merge of #109105 - compiler-errors:late-ct-in-anon-ct, r=oli-obk

Don't ICE for late-bound consts across `AnonConstBoundary`

Fixes #108194

matthiaskrgr committed:
Rollup merge of #109110 - compiler-errors:impossible-impl-mono, r=jackh726

Don't codegen impossible to satisfy impls

Fixes #109098

matthiaskrgr committed:
Rollup merge of #109116 - MaciejWas:add-modifies-receiver-diagn-when-method-not-found, r=petrochenkov

Emit diagnostic when calling methods on the unit type in method chains

Fixes #104204.

What this PR does: If a method is not found somewhere in a call chain, we check if we called earlier a method with signature `(&mut T, ...) -> ()`. If this is the case then we emit a diagnostic message.

For example given input:

```
vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
```

the current output is:
```
error[E0599]: no method named `sort` found for unit type `()` in the current scope
 --> hello.rs:3:72
  |
3 |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
  |                                                                        ^^^^ method not found in `()`

```

after this PR it will be:
```
error[E0599]: no method named `sort` found for unit type `()` in the current scope
 --> ./hello.rs:3:72
  |
3 |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
  |                                                                        ^^^^ method not found in `()`
  |

note: method `sort_by_key` modifies its receiver in-place, it is not meant to be used in method chains.
 --> ./hello.rs:3:53
  |
3 |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
  |                                                     ^^^^^^^^^^^ this call modifies its receiver in-place
```

bors committed:
Auto merge of #109130 - matthiaskrgr:rollup-dm3jza6, r=matthiaskrgr

Rollup of 9 pull requests

Successful merges:

 - #108722 (Support for Fuchsia RISC-V target)
 - #108880 (Remove tests/ui/impl-trait/in-trait/new-lowering-strategy in favor of using revisions on existing tests)
 - #108909 (Fix object safety checks for new RPITITs)
 - #108915 (Remove some direct calls to local_def_id_to_hir_id on diagnostics)
 - #108923 (Make fns from other crates with RPITIT work for -Zlower-impl-trait-in-trait-to-assoc-ty)
 - #109101 (Fall back to old metadata computation when type references errors)
 - #109105 (Don't ICE for late-bound consts across `AnonConstBoundary`)
 - #109110 (Don't codegen impossible to satisfy impls)
 - #109116 (Emit diagnostic when calling methods on the unit type in method chains)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

compiler-errors committed:
Simplify proc macro signature validity check

compiler-errors committed:
Tighter spans

rohaquinlop committed:
Rename 'src/bootstrap/native.rs' to llvm.rs

Renamed 'native.rs' to 'llvm.rs', also moved `TestHelpers` to `test.rs`.Replaced all the `native.rs` ocurrences at `src/bootstrap` files to `llvm.rs`

rohaquinlop committed:
Rename 'src/bootstrap/native.rs' to llvm.rs

Renamed 'native.rs' to 'llvm.rs', also moved `TestHelpers` to `test.rs`.Replaced all the `native.rs` ocurrences at `src/bootstrap` files to `llvm.rs`

weihanglo committed:
Update cargo

14 commits in 7d3033d2e59383fd76193daf9423c3d141972a7d..4a3c588b1f0a8e2dc8dd8789dbf3b6a71b02ed49
2023-03-08 17:05:08 +0000 to 2023-03-14 14:05:36 +0000
- ci: make clean-test-output a script for reuse (rust-lang/cargo#11848)
- Accurately show status when downgrading dependencies (rust-lang/cargo#11839)
- docs(contrib): Move Design Principles earlier in the book (rust-lang/cargo#11842)
- docs(contrib): Point compilation docs to doc comments (rust-lang/cargo#11841)
- `cargo install --git` multiple packages with binaries found hint (rust-lang/cargo#11835)
- Disable flaky auth tests when `gitoxide` runs them (rust-lang/cargo#11830)
- Add some documentation on writing cross-compilation tests (rust-lang/cargo#11825)
- chore: Use sparse protocol on stable CI (rust-lang/cargo#11829)
- Notice for potential unexpected shell expansions in help text of `cargo-add` (rust-lang/cargo#11826)
- Add tracking issue to gitoxide unstable docs (rust-lang/cargo#11822)
- Bump crates-io to 0.36.0 (rust-lang/cargo#11820)
- Bump to 0.71.0; update changelog (rust-lang/cargo#11815)
- docs(contrib): Move overview to lib (rust-lang/cargo#11809)
- Fix semver check for 1.68 (rust-lang/cargo#11817)

chenyukang committed:
Fix #109152, fix the scenario that we may can not get span of func

bors committed:
Auto merge of #107376 - aliemjay:remove-givens, r=lcnr

remove obsolete `givens` from regionck

Fixes #106567

r? `@lcnr` (feel free to reassign)

scottmcm committed:
Improved implementation and comments after code review feedback

scottmcm committed:
Split the mem-replace codegen test

Apparently in CI it's getting generated in the opposite order, one function per file will make the test pass either way.

bors committed:
Auto merge of #109089 - compiler-errors:opt_rpitit_info-follow-up, r=spastorino

Encode `opt_rpitit_info` for associated types

Follow-up, only last commit matters

r? `@spastorino`

This needs a perf run after the parent pr lands

michaelwoerister committed:
Remove mw from review rotation.

bors committed:
Auto merge of #109035 - scottmcm:ptr-read-should-know-undef, r=WaffleLapkin,JakobDegen

Ensure `ptr::read` gets all the same LLVM `load` metadata that dereferencing does

I was looking into `array::IntoIter` optimization, and noticed that it wasn't annotating the loads with `noundef` for simple things like `array::IntoIter<i32, N>`.  Trying to narrow it down, it seems that was because `MaybeUninit::assume_init_read` isn't marking the load as initialized (<https://rust.godbolt.org/z/Mxd8TPTnv>), which is unfortunate since that's basically its reason to exist.

The root cause is that `ptr::read` is currently implemented via the *untyped* `copy_nonoverlapping`, and thus the `load` doesn't get any type-aware metadata: no `noundef`, no `!range`.  This PR solves that by lowering `ptr::read(p)` to `copy *p` in MIR, for which the backends already do the right thing.

Fortuitiously, this also improves the IR we give to LLVM for things like `mem::replace`, and fixes a couple of long-standing bugs where `ptr::read` on `Copy` types was worse than `*`ing them.

Zulip conversation: <https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Move.20array.3A.3AIntoIter.20to.20ManuallyDrop/near/341189936>

cc `@erikdesjardins` `@JakobDegen` `@workingjubilee` `@the8472`

Fixes #106369
Fixes #73258

Dylan-DPC committed:
Rollup merge of #109109 - compiler-errors:polymorphize-foreign, r=Nilstrieb

Use `unused_generic_params` from crate metadata

Due to the way that `separate_provide_extern` interacted with the implementation of `<ty::InstanceDef<'tcx> as Key>::query_crate_is_local`, we actually never hit the foreign provider for `unused_generic_params`.

Additionally, since the *local* provider of `unused_generic_params` calls `should_polymorphize`, which always returns false if the def-id is foreign, this means that we never actually polymorphize monomorphic instances originating from foreign crates.

We don't actually encode `unused_generic_params` for items where all generics are used, so I had to tweak the foreign provider to fall back to `ty::UnusedGenericParams::new_all_used()` to avoid more ICEs when the above bugs were fixed.

Dylan-DPC committed:
Rollup merge of #108991 - ozkanonur:new-llvm-flag, r=albertlarsan68

add `enable-warnings` flag for llvm, and disable it by default.

This flag allows to turn off warnings of llvm compilation for people who are not interested on those warnings.

Dylan-DPC committed:
Rollup merge of #109136 - compiler-errors:simplify-proc-macro-checking, r=oli-obk

Simplify proc macro signature validity check

Use an `ObligationCtxt` instead of `normalize_erasing_regions` + `DeepRejectCtxt`. This should both give us a more accurate error message, and also avoid issues like not-well-formed proc macro signatures. Also, let's fall back on the regular type mismatch error reporting for making these diagnostic notes, instead of hard-coding a bunch of specific diagnostics.

Fixes #109129

Dylan-DPC committed:
Rollup merge of #109111 - MU001999:master, r=jyn514

Create dirs for build_triple

Fixes #109103

Dylan-DPC committed:
Rollup merge of #109154 - chenyukang:yukang/fix-109152, r=compiler-errors

Fix MappingToUnit  to support no span of arg_ty

Fixes #109152

Dylan-DPC committed:
Rollup merge of #109150 - weihanglo:update-cargo, r=weihanglo

Update cargo

14 commits in 7d3033d2e59383fd76193daf9423c3d141972a7d..4a3c588b1f0a8e2dc8dd8789dbf3b6a71b02ed49
2023-03-08 17:05:08 +0000 to 2023-03-14 14:05:36 +0000
- ci: make clean-test-output a script for reuse (rust-lang/cargo#11848)
- Accurately show status when downgrading dependencies (rust-lang/cargo#11839)
- docs(contrib): Move Design Principles earlier in the book (rust-lang/cargo#11842)
- docs(contrib): Point compilation docs to doc comments (rust-lang/cargo#11841)
- `cargo install --git` multiple packages with binaries found hint (rust-lang/cargo#11835)
- Disable flaky auth tests when `gitoxide` runs them (rust-lang/cargo#11830)
- Add some documentation on writing cross-compilation tests (rust-lang/cargo#11825)
- chore: Use sparse protocol on stable CI (rust-lang/cargo#11829)
- Notice for potential unexpected shell expansions in help text of `cargo-add` (rust-lang/cargo#11826)
- Add tracking issue to gitoxide unstable docs (rust-lang/cargo#11822)
- Bump crates-io to 0.36.0 (rust-lang/cargo#11820)
- Bump to 0.71.0; update changelog (rust-lang/cargo#11815)
- docs(contrib): Move overview to lib (rust-lang/cargo#11809)
- Fix semver check for 1.68 (rust-lang/cargo#11817)

r? `@ghost`

Dylan-DPC committed:
Rollup merge of #109157 - michaelwoerister:tb, r=Mark-Simulacrum

Remove mw from review rotation for a while

I'm triple booked right now 📚

bjorn3 committed:
Use patched git-subtree from bjorn3/git@tqc-subtree-portable

This patched has been necessary for subtree syncs from the start, but
previously it was necessary to locally install tqc's patched git
version, which hasn't been updated for quite a while. I made a small
change to allow downloading it as script without requiring an entire git
installation for the patched version.

bjorn3 committed:
Introduce Box::new in mini_core

bjorn3 committed:
Sync from rust 1716932743a7b3705cbf0c34db0c4e070ed1930d

bjorn3 committed:
Rustup to rustc 1.70.0-nightly (171693274 2023-03-14)

bjorn3 committed:
Merge commit 'dec0daa8f6d0a0e1c702f169abb6bf3eee198c67' into sync_cg_clif-2023-03-15

bors committed:
Auto merge of #109164 - Dylan-DPC:rollup-0bwxwos, r=Dylan-DPC

Rollup of 7 pull requests

Successful merges:

 - #108991 (add `enable-warnings` flag for llvm, and disable it by default.)
 - #109109 (Use `unused_generic_params` from crate metadata)
 - #109111 (Create dirs for build_triple)
 - #109136 (Simplify proc macro signature validity check)
 - #109150 (Update cargo)
 - #109154 (Fix MappingToUnit  to support no span of arg_ty)
 - #109157 (Remove mw from review rotation for a while)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

lqd committed:
Revert "Auto merge of #107376 - aliemjay:remove-givens, r=lcnr"

This reverts commit e84e5ff04a647ce28540300244a26ba120642eea, reversing
changes made to 1716932743a7b3705cbf0c34db0c4e070ed1930d.

bors committed:
Auto merge of #109169 - bjorn3:sync_cg_clif-2023-03-15, r=bjorn3

Sync rustc_codegen_cranelift

Bunch of bug fixes this time. Also an update to Cranelift 0.93 which adds a brand new optimization pass which cg_clif exposes when using `--release`. And various improvements to cg_clif's test suite, making it faster to run. And finally two small perf improvements.

r? `@ghost`

`@rustbot` label +A-codegen +A-cranelift +T-compiler

bors commented:
:umbrella: The latest upstream changes (presumably #109164) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors committed:
Auto merge of #108282 - cjgillot:mir-checked-sh, r=tmiasko

Implement checked Shl/Shr at MIR building.

This does not require any special handling by codegen backends,
as the overflow behaviour is entirely determined by the rhs (shift amount).

This allows MIR ConstProp to remove the overflow check for constant shifts.

~There is an existing different behaviour between cg_llvm and cg_clif (cc `@bjorn3).`
I took cg_llvm's one as reference: overflow if `rhs < 0 || rhs > number_of_bits_in_lhs_ty`.~

EDIT: `cg_llvm` and `cg_clif` implement the overflow check differently. This PR uses `cg_llvm`'s implementation based on a `BitAnd` instead of `cg_clif`'s one based on an unsigned comparison.

bors committed:
Auto merge of #109183 - lqd:revert-107376, r=compiler-errors

Revert #107376 to fix potential `bincode` breakage and `rustc-perf` benchmark.

#107376 caused `rustc-perf`'s `webrender` benchmark to break, by regressing on the `bincode-1.3.3` crate.

~~This PR is a draft revert in case we can't land a fix soon enough, and we'd like to land the revert instead~~

(Though I myself think it'd be safer to do the revert, and run crater when relanding #107376.)

cc `@aliemjay`

bors committed:
Auto merge of #108809 - lqd:fix-ignore, r=pietroalbini

fix ignore header in MSVC test

From `@pietroalbini's` [zulip message](https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/better.20compiletest.20ignore.20messages/near/339845864)

> there are tests like `tests/ui/panic-runtime/unwind-tables-target-required.rs` which have `only-x86_64-windows-msvc` which I'm pretty sure is invalid

This test is currently ignored on x64 MSVC CI because of this incorrect target. We'll see if it still passes.

r? `@pietroalbini`

rohaquinlop committed:
Merge branch 'rename-native.rs-to-llvm.rs' of https://github.com/rohaquinlop/rust into rename-native.rs-to-llvm.rs

Mark-Simulacrum commented:
This seems to have included a merge or similar, you probably want to rebase. @rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #109384) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
